src/include/execution/exec/output.h

```diff
@@ -45,23 +45,50 @@ class EXPORT OutputBuffer {
    */
   explicit OutputBuffer(sql::MemoryPool *memory_pool, uint16_t num_cols, uint32_t tuple_size, OutputCallback callback)
       : memory_pool_(memory_pool),
-        num_tuples_(0),
         tuple_size_(tuple_size),
-        tuples_(
-            reinterpret_cast<byte *>(memory_pool->AllocateAligned(BATCH_SIZE * tuple_size, alignof(uint64_t), true))),
-        callback_(std::move(callback)) {}
+        id_(0),
+        callback_(std::move(callback)) {
+          max_thread_ = std::thread::hardware_concurrency();
+          if (max_thread_ == 0) {
+            // Single thread if fail to get the number of cores
+            max_thread_ = 1;
+          }
+          num_tuples_ = reinterpret_cast<uint32_t *>(
+          memory_pool->AllocateAligned(max_thread_ * sizeof(uint32_t), alignof(uint32_t), true));
+        }
+
+  /**
+   * Insert the corresponding buffer of a thread to the map if not created yet
+   * @param id The thread ID
+   */
+  void InsertIfAbsent(std::thread::id id) {
+    if (buffer_map_.find(id) == buffer_map_.end()) {
```

Comment:
I believe there is a race condition between here and `AllocOutputSlot` if we're running inside a thread pool or something where thread ids are reused?

Reply:
Could you explain more about the race condition? I think threads could be reused across different output buffers, but then they would be different instances of output buffers. Buffer map itself is a concurrent map, which guarantees thread-safety for one output buffer.